### ISSUE TITLE
Gauge double registration warning for Kafka consumer metrics

### ISSUE THREAD

**Issue by MichaelVilser** [micrometer-metrics/micrometer#5757]
Since https://github.com/micrometer-metrics/micrometer/pull/5688 we observe warnings being logged for the default kafka consumer gauge metrics such as 'kafka.consumer.fetch.manager.bytes.consumed.rate' in the form of "This Gauge has been already registered".

**Stack trace of when a gauge metric is registered**
[1stTime_on_startup.txt](https://github.com/user-attachments/files/18238005/1stTime_on_startup.txt)
[2ndTime_after_startup.txt](https://github.com/user-attachments/files/18238007/2ndTime_after_startup.txt)

Looking at this function, it seems like this behavior is actually intended:
https://github.com/micrometer-metrics/micrometer/blob/9d650ca3634ce11b70c18abf700252a7ab4a27c1/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/kafka/KafkaMetrics.java#L138-L147

**Comment by oliver-schwab** [comment]
We also see these warnings in our applications, but only in the ones with multiple consumers. Is this also intended?

**Comment by jonatan-ivanov** [comment]
Thanks for the issue!
`KafkaMetrics` needs to re-register `Meter`s since the Kafka client pulls the rug and drops `KafkaMetric` instances (where it records the data) in some cases. Re-registration is expected but meters should be removed first so the double-registration might be a bug.

A proper way to "fix" this would be moving the instrumentation into the Kafka Client itself, so we can instrument this natively and don't need to wrestle with the Kafka Client. Please comment this if you would like to see that happen: https://issues.apache.org/jira/browse/KAFKA-15191

Normally I would say that we have a bug and we don't remove something but based on your first attached file ([1stTime_on_startup.txt](https://github.com/user-attachments/files/18238005/1stTime_on_startup.txt)) it seems this happens at startup so it is also possible that Kafka publishes something twice?

Does any of you have a minimal sample project to reproduce this issue?
Does this happen every time at startup or occasionally?
Does this happen only in case of `kafka.consumer.fetch.manager.bytes.consumed.rate` or other metrics are also affected?

**Comment by tzie0062** [comment]
I am seeing the same issue for all of my (spring-)kafka apps.

@jonatan-ivanov here's a [minimal app](https://github.com/tzie0062/spring-kafka-micrometer-registration-issue) that reproduces the issue

* So far I only observed this directly after startup/deployment
* The metric being logged changes (so it does not seem to be related to a single one)

**Comment by jonatan-ivanov** [comment]
>I also realized that while we added logging for gauge double registration, any of the "functional" meters are affected - FunctionCounter, FunctionTimer. That's relevant to KafkaMetrics because it only registers Gauges and FunctionCounters. /cc @jonatan-ivanov

Good point, I opened https://github.com/micrometer-metrics/micrometer/pull/6070.

**Comment by shakuzen** [comment]
Could those who were seeing this issue please try out snapshots to confirm the fix we merged works for you? It is in 1.13.13-SNAPSHOT, 1.14.6-SNAPSHOT, and 1.15.0-SNAPSHOT.

**Comment by thuri** [comment]
tested this with 1.4.6-SNAPSHOT against our application and it looks ok to me. 
The warning no longer occurs.
I had a quick look at some metrics and think they're looking good. I have some NaN values delivered by prometheus actuator for kafka_consumer_fetch_manager_records_lag but that may not be an issue because of this fix. Tested this only on a low traffic deployment

**Comment by a-dudko** [comment]
We migrated from io.micrometer:micrometer-core:jar:1.13.4 to io.micrometer:micrometer-core:jar:1.15.1 and strangely enough started getting the aforementioned warning:

`This Gauge has been already registered (MeterId{name='kafka.consumer.connection.count', tags=[tag(client.id=577c048e-6a7e-4513-9da3-d511b3fd431a-StreamThread-3-consumer)]}), the registration will be ignored. Note that subsequent logs will be logged at debug level.`

**Comment by jonatan-ivanov** [comment]
@a-dudko Thanks for letting us know!

Does this happen only in case of kafka.consumer.fetch.manager.bytes.consumed.rate or other metrics are also affected?

Could you please open a new issue for this and provide a minimal sample Java project to reproduce this issue so we can more easily investigate and ensure any fix is working properly for your use case?

**Comment by jens-brimfors-wolt** [comment]
Looks like #6068 was merged into a 1.13.x fixbranch, but never ended up in main..
https://github.com/shakuzen/micrometer/blob/main/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/kafka/KafkaMetrics.java#L185
Maybe that's why i see this in 1.14.x versions now 🤦
@shakuzen What do you think?

**Comment by jonatan-ivanov** [comment]
Hi @jens-brimfors-wolt,

I'm not @shakuzen but I think you linked (/looking at) his fork of Micrometer instead of Micrometer itself. This is the file in the repo we release: [`micrometer-metrics/micrometer/blob/main/micrometer-core/.../KafkaMetrics.java`](https://github.com/micrometer-metrics/micrometer/blob/main/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/kafka/KafkaMetrics.java)
This is the commit where the fix was merged into `main`: https://github.com/micrometer-metrics/micrometer/commit/1893e65ec9e26ba49fa6df53a776913b8ab6119a

Please either look at the official repo ([`micrometer-metrics/micrometer`](https://github.com/micrometer-metrics/micrometer)) or download the sources jar from Maven Central with your IDE and look at the code there.
